### PR TITLE
Ensure `mlflow.pyfunc.spark_udf` supports  `ArrayType(StructType))`

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Ensures `mlflow.pyfunc.spark_udf` supports `ArrayType(StructType))`.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/pyfunc/test_spark.py [hunk 1]
```diff
@@ -723,3 +726,39 @@ def test_spark_udf_stdin_scoring_server(spark, monkeypatch):
         df = spark.createDataFrame(X)
         result = df.select(udf(*X.columns)).toPandas()
         np.testing.assert_almost_equal(result.to_numpy().squeeze(), model.predict(X))
+
+
+@pytest.mark.skipif(
+    Version(pyspark.__version__) < Version("3.4.0"), reason="requires spark >= 3.4.0"
+)
```
Comment (harupy) [inline tests/pyfunc/test_spark.py:733]:
```suggestion
# @pytest.mark.skipif(
#    Version(pyspark.__version__) < Version("3.4.0"), reason="requires spark >= 3.4.0"
# )
```

File: tests/pyfunc/test_spark.py [hunk 2]
```diff
@@ -723,3 +726,39 @@ def test_spark_udf_stdin_scoring_server(spark, monkeypatch):
         df = spark.createDataFrame(X)
         result = df.select(udf(*X.columns)).toPandas()
         np.testing.assert_almost_equal(result.to_numpy().squeeze(), model.predict(X))
+
+
+# @pytest.mark.skipif(
+#    Version(pyspark.__version__) < Version("3.4.0"), reason="requires spark >= 3.4.0"
+# )
```
Comment (harupy) [inline tests/pyfunc/test_spark.py:733]: Just for testing

Comment (harupy) [reply tests/pyfunc/test_spark.py:733]: Confirmed: https://github.com/mlflow/mlflow/actions/runs/4538225945/jobs/7996927410

Comment (harupy) [inline tests/pyfunc/test_spark.py:733]:
```suggestion
@pytest.mark.skipif(
   Version(pyspark.__version__) < Version("3.4.0"), reason="requires spark >= 3.4.0"
)
```

**Comment by harupy** [comment]
@mlflow-automation autoformat

File: tests/pyfunc/test_spark.py (hunk 1 above)
Comment (harupy) [inline tests/pyfunc/test_spark.py:733]:
```suggestion
# TODO: Remove `skipif` once pyspark 3.4.0 is released
@pytest.mark.skipif(
    Version(pyspark.__version__) < Version("3.4.0"), reason="requires spark >= 3.4.0"
)
```

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -121,6 +121,10 @@ jobs:
         run: |
           source dev/setup-ssh.sh
           ./dev/run-python-tests.sh
+      - name: Run tests for pyspark 3.4.0
+        run: |
+          pip install https://dist.apache.org/repos/dist/dev/spark/v3.4.0-rc1-bin/pyspark-3.4.0.tar.gz
+          pytest tests/pyfunc/test_spark.py::test_spark_udf_array_of_structs
```
Comment (harupy) [inline .github/workflows/master.yml:127]:
```suggestion
      # TODO: Remove this step once pyspark 3.4 is released
      - name: Run tests for pyspark 3.4.0
        run: |
          pip install https://dist.apache.org/repos/dist/dev/spark/v3.4.0-rc1-bin/pyspark-3.4.0.tar.gz
          pytest tests/pyfunc/test_spark.py::test_spark_udf_array_of_structs
```

**Review by WeichenXu123 — APPROVED** [review]

**Comment by WeichenXu123** [comment]
The PR title needs to be updated to be consistent with PR content.

**Comment by harupy** [comment]
@WeichenXu123 updated the title.

**Pull request merged**